ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,295 @@
+# Architectural Decision Record 003: PyBind11 for C++ bindings
+
+## Status
+[**Proposed** | <s>Accepted</s> | <s>Deprecated</s> | <s>Superseded by [ADR-XXX]</s>]
+
+## Last Updated
+2025-09-18
+
+## Decision
+
+The decision is to use Pybind11 for C++ bindings (see [analysis](#analysis)). 
+Developers should follow the [design guidline](#design-guidlines).
+
+
+## Context
+
+ECMWF maintains a suite of high-performance C++ libraries that are increasingly
+required to be accessible from Python — for both internal workflows and external
+users. Historically, this integration has been achieved using CFFI, which
+requires exposing a C-compatible API and writing additional Python wrappers.
+
+However, this approach presents several technical and business challenges:
+
+
+### Technical Challenges
+
+- Limited C++ Support: CFFI does not natively support C++ features such as
+  classes, templates, or STL containers. This forces developers to write and
+  maintain intermediate C wrappers, which is error-prone and hard to scale.
+- High Maintenance Overhead: The need to manually manage memory, object
```

Comment:
Need to mention idiomatic differences. C APIs don't work well with RAII. Python has a very different view of object lifetimes to C or C++. Wiring through both of those changes is challenging.

Reply:
Reformulated this part an emphasized the procedural nature of CFFI bindings. Also emphasized the required mitigation of different approaches to obj lifetime.